doc/dev/kdb-operations.md

```diff
@@ -0,0 +1,159 @@
+
+- The `get` operation has: `init`, `resolver`, `cachecheck`, `prestorage`, `storage` and `poststorage`.
+- The `set` operation has: `resolver`, `prestorage`, `storage` and `poststorage` followed by `precommit`, `commit` and `postcommit` if the previous phases where successful or by `prerollback`, `rollback` and `postrollback` if the previous phases failed.
+
+These phases are implemented by a backend plugin.
+Read the [Documentation on Backend Plugins](./backend-plugins.md) for more information on what backend plugins do.
+
+## `open` Operation
+
+The `open` operation implemented in `kdbOpen` is the first thing that happens to all `KDB` instances.
+
+The basic flow of this operation is:
+
+1. Create empty `KDB` instance
+2. Configure `KDB` instance for bootstrap
+3. Run bootstrap `get` operation
+4. Mount global plugins
+5. Process contract
+6. Parse mountpoints
+7. Configure `KDB` instance with real mountpoints
+8. Add hardcoded mountpoints to `KDB` instance
+
+Namespaces in mountpoint configs:
+
+- cascading mountpoints act as an alias for identical `proc:/`, `dir:/`, `user:/`, `system:/` and `default:/` mountpoints.
+- `meta:/` mountpoints are illegal
+- explicit `default:/` mountpoints are illegal
+  They can only exist via cascading mountpoints.
+  This is so that default values can be checked and validated by plugins.
+- `spec:/` mountpoints may exist, but they are treated differently during `get` and `set`
+- `proc:/` mountpoints may exist, but they are read-only (see below)
+- `dir:/`, `user:/` and `system:/` mountpoints can be created without restrictions
+
+Other restrictions:
+
+- Creating mountpoints for root keys (`system:/`, `user:/`, etc.) is not allowed.
+  These parts of the KDB always use hardcoded backends that can only be configured at compile-time.
+- Creating a mountpoint for `system:/elektra` or any mountpoint below is not allowed.
+  This part of the KDB will always be stored in the bootstrap backend.
+  While the only part that is actually required for the bootstrap process is the one below `system:/elektra/mountpoints`, everything below `system:/elektra` is reserved and should only be used internally.
+  Additionally, `system:/elektra/version` and `system:/elektra/modules` will always point to hardcoded read-only backends containing information about this Elektra installation.
+
+## `get` Operation
+
+The purpose of the `get` operation is to read data stored in backends into a `KDB` instance.
+
+> **Note:** Some details of a `get` operation are defined in the [contract with backend plugins](./backend-plugins.md).
+
+Properties of `kdbGet()`:
+
+- After calling `kdbGet (kdb, ks, parentKey)`, the KeySet `ks` will contain _all keys_ (including their values) that are stored in _any backend_ with a mountpoint that is _below `parentKey`_.
+- After calling `kdbGet (kdb, ks, parentKey)`, below `parentKey` the KeySet `ks` will _only_ contain keys that are stored in a backend.
```

Comment:
No, it does not. The first statement says, all config below `parentKey` will be read fully. The second statement says, there won't be anything other than the config that was read below `parentKey`.

Basically, the stuff not below `parentKey` will be ignored. But below `parentKey` the `KeySet` will represent the real KDB data. It more or less means that `kdbGet(kdb, ks, parentKey)` starts with a `ksDel (ksCut (ks, parentKey))`. We cannot keep anything below `parentKey`, because then this wouldn't work:

1. Call `kdbGet (kdb, ks, parentKey)`.
2. Externally remove a key below `parentKey`.
3. Call `kdbGet (kdb, ks, parentKey)` again to get the fresh data.

If we keep keys below `parentKey` that already exist, we won't know that a key was removed.

Reply:
Are you talking about calling `kdbGet (kdb, ks, parentKey)` **again**?

To make it clear for me: What exactly do you want to change compared to the current situation?

Reply:
> What exactly do you want to change compared to the current situation?

Hopefully nothing, otherwise I have misunderstood the current situation or the current code is broken.

---

It is not currently described like this in the document, because I haven't found a way to formulate this concisely and understandably yet.

I'll give you an example of how it should ideally work, and then I'll describe what the document says based on the example.

```c
Key * parentKey = keyNew ("/foo", KEY_END);
KeySet * ks = ksNew (4,
                     keyNew ("user:/foo/bar", KEY_VALUE, "1", KEY_END),
                     keyNew ("user:/foo/baz/bar", KEY_VALUE, "2", KEY_END),
                     keyNew ("default:/foo/baz/bar", KEY_VALUE, "3", KEY_END),
                     keyNew ("user:/bar/foo", KEY_VALUE, "4", KEY_END),
                     KS_END
                    );

kdbGet (kdb, ks, parentKey);
```

In this example, we assume that a backend is mounted at `user:/foo/baz`. There are no other backends mounted below `/foo`.

After executing this snippet:
1. `user:/bar/foo` will be in `ks` and it's value will be `4`.
  The key `user:/bar/foo` is completely untouched, because it is not below `/foo`.
2. `user:/foo/bar` will be in `ks` and it's value will be `1`.
   The key `user:/foo/bar` is completely untouched, because it is not below the mountpoint of any of the backends below `/foo`.
3. `user:/foo/baz/bar` may or may not be in `ks` and it's value may or may not be `2`.
   The key `user:/foo/baz/bar` is below the mountpoint `user:/foo/baz`. If there is a value for this key in the backend (which might coincidentally also be `2`), then `user:/foo/baz/bar` will be replaced appropriately. If there is no value for this key, then it will be removed.
   In any case, the `Key` instance created in 4 of the snippet will always be deleted by `kdbGet`.
4. `default:/foo/baz/bar` will be in `ks` and it's value may or may not be `4`.
  The key `default:/foo/baz/bar` will be completely untouched, if `spec:/foo/baz/bar` doesn't exist or doesn't define a `default`. If there is a `default` in `spec:/foo/baz/bar`, then `default:/foo/baz/bar` will be replaced appropriately.

This mode of operation guarantees that `ks` will always reflect the data stored in backends, where such data could exist. At the special handling  exception for `default:/` keys.